src/ucp/wireup/wireup_cm.c

```diff
@@ -1272,11 +1271,10 @@ ucp_ep_cm_server_create_connected(ucp_worker_h worker, unsigned ep_init_flags,
         goto err_destroy_ep;
     }
 
-    cm_wireup_ep = ucp_ep_get_cm_wireup_ep(ep);
-    ucs_assert(cm_wireup_ep != NULL);
-    ucp_wireup_dst_rsc_indices_fill(ep, &ucp_ep_config(ep)->key, remote_addr,
-                                    addr_indices,
-                                    cm_wireup_ep->dst_rsc_indices);
+    /* Since CM lane isn't available and no previous configuration available on
+     * the endpoint, save peer's resource indices after creating CM lane */
```

Comment:
it means the previously saved `cm_wireup_ep->dst_rsc_indices` would not be valid prior to calling `ucp_wireup_check_config_intersect`?
why is this not needed also for client code?

Reply:
> it means the previously saved cm_wireup_ep->dst_rsc_indices would not be valid prior to calling ucp_wireup_check_config_intersect?

there are no previously saved `cm_wireup_ep->dst_rsc_indices` at this point, all indices are set to `UCP_NULL_RESOURCE`

> why is this not needed also for client code?

client creates CM lane, connects to the server and only after that it gets server's CM private data and `ucp_wireup_check_config_intersect` saves it to `cm_wireup_ep->dst_rsc_indices`

Reply:
for client side: seems ucp_wireup_check_config_intersect compares previous cm_wireup_ep->dst_rsc_indices to the new ones. in this case, AFAIK the previous ones will be all UCP_NULL_RESOURCE which is ok?

Reply:
> for client side: seems ucp_wireup_check_config_intersect compares previous cm_wireup_ep->dst_rsc_indices to the new ones. in this case, AFAIK the previous ones will be all UCP_NULL_RESOURCE which is ok?

yes, it is ok for the first iteration, we don't expect reconfiguration.
but then we will have updated `cm_wireup_ep->dst_rsc_indices` which will be used during WIREUP_MSG phase to find best configuration

Reply:
> so maybe the problem is the server does not call ucp_wireup_check_config_intersect (like the client)?

it calls, but before CM lane is created.

Reply:
if we will create CM lane and save it to UCP EP's `uct_eps` before real configuration exists.
actually, client and server behaviors are aligned:
Client/Server creates UCP EP, does `ucp_wireup_init_lanes` whichs call `ucp_wireup_check_config_intersect ` - but it is just called, doesn't do any real work, since no CM lane and previously create configuration and then create CM lane

Reply:
the only difference is that when creating CM lane on sever, we already know destination resource indices and we set them after CM lane is created.
these indices will be used during next `ucp_wireup_init_lanes` and `ucp_wireup_check_config_intersect`

Reply:
moved `ucp_wireup_dst_rsc_indices_save` to `ucp_wireup_ep_create`

Reply:
@yosefe is it what you meant?